### ISSUE TITLE
Preload option in RealGalaxyCatalog only preload HDUs not data

### ISSUE THREAD

**Issue by EiffL** [GalSim-developers/GalSim#884]
The preload option in RealGalaxyCatalog, and by extension, COSMOSCatalog does open all the files and reads in the HDUs but the data is not actually loaded into RAM. From what I can read in issue #168 this is by design, but I have a use case where I actually need to have the whole catalog loaded in memory (I need to very rapidly produce galaxy images in random order).
 
I can manually force the catalog to preload all the data with the following:
```
# Doesn't actually preload, returns very quickly
cat = galsim.COSMOSCatalog(preload=True)

# Takes a while as it cycles through the files, memory usage steadily increases in the process
for f in cat.real_cat.loaded_files:
    for h in cat.real_cat.loaded_files[f]:
        h.data;
```

But what to do in this situation ? Should there be separate preload and preload_HDU options ? 

If you need to preload the HDUs, most likely you are going to loop through the entire catalog which will progressively load the whole thing in memory anyway, so in terms of memory usage it doesn't change much. But preloading the catalog by sequentially reading all the files is much faster than randomly accessing different HDUs until the whole catalog is loaded.

**Comment by rmjarvis** [comment]
I'm fine with just making the API change to have preload actually load in all the data as you had expected it to do.  (Although I think the doc string describes the current behavior accurately.)

The case for preloading as we currently do it has largely gone away, I think.  Some time ago I did some timing tests with and without preloading, and there was no longer the kind of difference we were seeing back in 2012 when we were working on issue #168.  If you use all the galaxies, the times were the same, and if you didn't, preloading was slower.  So we don't promote `preload` anywhere in the demos, and I don't ever bother with it myself.

So I'd be fine with changing it to actually load in all the data, rather than just the header info, as the word "preload" sounds like it would be doing.

**Comment by EiffL** [comment]
Ok, great. I'm happy to make a pull request. Is the solution I quoted in the issue  working for you or can you think of a smarter way to do this ?

**Comment by rmjarvis** [comment]
Looks right to me.  You can look in the astropy docs to see if there is something simpler, but what you wrote does what you want and looks efficient enough.

**Comment by rmjarvis** [comment]
Merged via #885
